### PR TITLE
Add IPBasicAuth middleware

### PR DESCRIPTION
**Pull request by sashah2**

### CODE REVIEW

File: httpmw/ip_basic_auth.go [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+	"bytes"
+	"encoding/base64"
+	"github.com/gin-gonic/gin"
+	"net"
+	"net/http"
+	"strings"
+)
+
+// Range define range with start and end IPs
+type Range struct {
```
Comment (protsack-stephan) [inline httpmw/ip_basic_auth.go]: Does this type needs to be public? Looks like it's not going to be used outside this package.

File: httpmw/ip_basic_auth.go [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+	Start net.IP
+	End   net.IP
```
Comment (protsack-stephan) [inline httpmw/ip_basic_auth.go]: Looks like those properties should private as well.

File: httpmw/ip_basic_auth.go [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+
+func (a authPairs) searchCredential(authValue string) (string, bool) {
+	if authValue == "" {
+		return "", false
+	}
+	for _, pair := range a {
+		if pair.value == authValue {
+			return pair.user, true
+		}
+	}
```
Comment (protsack-stephan) [inline httpmw/ip_basic_auth.go:32]: What about something like this:

```go
if authValue != "" {
	for _, pair := range a {
		if pair.value == authValue {
			return pair.user, true
		}
	}
}
	
```

File: httpmw/ip_basic_auth.go [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+		if user == "" {
+			continue
+		}
+
+		value := authorizationHeader(user, password)
+		pairs = append(pairs, authPair{
+			value: value,
+			user:  user,
+		})
+	}
```
Comment (protsack-stephan) [inline httpmw/ip_basic_auth.go:108]: Probably need a white-space after this.

File: httpmw/ip_basic_auth.go [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+	return pairs
+}
+
+func authorizationHeader(user, password string) string {
+	base := user + ":" + password
```
Comment (protsack-stephan) [inline httpmw/ip_basic_auth.go]: Can use `fmt` package here:

```go
fmt.Sprintf("%s:%s", user, password)
```

File: httpmw/ip_basic_auth.go [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+
+	return "Basic " + base64.StdEncoding.EncodeToString([]byte(base))
```
Comment (protsack-stephan) [inline httpmw/ip_basic_auth.go]: Can use `fmt` package:

```go
fmt.Sprintf("Basic %s" , base64.StdEncoding.EncodeToString([]byte(base)))
```

File: httpmw/ip_basic_auth.go [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+)
+
+// Range define range with start and end IPs
+type Range struct {
+	Start net.IP
+	End   net.IP
+}
+
+// Ranges array of Range
+type Ranges []Range
```
Comment (protsack-stephan) [inline httpmw/ip_basic_auth.go]: This should probably be private as well.

File: httpmw/ip_basic_auth.go [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+func checkIP(ipRange Range, ip string) bool {
+	input := net.ParseIP(ip)
+
+	return bytes.Compare(input, ipRange.Start) >= 0 && bytes.Compare(input, ipRange.End) <= 0
+}
+
+func processAccounts(accounts gin.Accounts) authPairs {
+	pairs := make(authPairs, 0, len(accounts))
+
+	if len(accounts) <= 0 {
```
Comment (protsack-stephan) [inline httpmw/ip_basic_auth.go]: We don't need this check, if we have `0` accounts loop won't iterate so this is not needed.

File: httpmw/ip_basic_auth.go [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+		return pairs
+	}
+
+	for user, password := range accounts {
+		if user == "" {
+			continue
+		}
+
+		value := authorizationHeader(user, password)
```
Comment (protsack-stephan) [inline httpmw/ip_basic_auth.go]: What about:
```go
if user != "" {
    pairs = append(pairs, authPair{
	value: authorizationHeader(user, password),
	user:  user,
    })
}
```

**Review by protsack-stephan — COMMENTED** [review]
Good job, just a couple nitpicks.

File: httpmw/ip_basic_auth.go (hunk 4 above)
Comment (sashah2) [reply httpmw/ip_basic_auth.go:108]: good catch!

**Pull request merged**